File: src/main/java/net/vpg/game2048/Move.java
package net.vpg.game2048;

public enum Move {
    UP   (0, 0, 0, 1),
    DOWN (1, 0, 0, 1),
    LEFT (0, 0, 1, 0),
    RIGHT(0, 1, 1, 0);

    final int row;
    final int column;
    final int rowChange;
    final int columnChange;

    Move(int row, int column, int rowChange, int columnChange) {
        this.row = row;
        this.column = column;
        this.rowChange = rowChange;
        this.columnChange = columnChange;
    }

    public static Move fromKey(char key) {
        return switch (key) {
            case 'u', 'U' -> UP;
            case 'd', 'D' -> DOWN;
            case 'l', 'L' -> LEFT;
            case 'r', 'R' -> RIGHT;
            default -> null;
        };
    }
}

File: src/main/java/net/vpg/game2048/Move.kt
package net.vpg.game2048

enum class Move(val row: Int, val column: Int, val rowChange: Int, val columnChange: Int) {
    UP   (0, 0, -1, 0),
    DOWN (1, 0, +1, 0),
    LEFT (0, 0, 0, -1),
    RIGHT(0, 1, 0, +1);
}
